execute/tokendata/observer.go

```diff
@@ -2,42 +2,145 @@ package tokendata
+		if c.USDCCCTPObserverConfig != nil {
+			observers[i] = usdc.NewUSDCCCTPTokenDataObserver(*c.USDCCCTPObserverConfig)
+		} else {
+			return nil, errors.New("unsupported token data observer")
+		}
+	}
+	return NewCompositeObservers(observers...), nil
+}
+
+func NewCompositeObservers(observers ...TokenDataObserver) *CompositeTokenDataObserver {
+	return &CompositeTokenDataObserver{observers: observers}
 }
 
+// Observe start with stubbing exectypes.TokenDataObservations with empty data based on the supported tokens.
+// Then it iterates over all observers and merges token data returned from them into the final result.
 func (t *CompositeTokenDataObserver) Observe(
 	ctx context.Context,
-	messages exectypes.MessageObservations,
+	msgObservations exectypes.MessageObservations,
 ) (exectypes.TokenDataObservations, error) {
-	var tokenObservation exectypes.TokenDataObservations
-	for _, ob := range t.Observers {
-		tokenData, err := ob.Observe(ctx, messages)
+	tokenDataObservations := t.initTokenDataObservations(msgObservations)
+
+	for _, ob := range t.observers {
+		tokenData, err := ob.Observe(ctx, msgObservations)
+		if err != nil {
+			return nil, err
+		}
+		tokenDataObservations, err = merge(tokenDataObservations, tokenData)
 		if err != nil {
 			return nil, err
 		}
-		tokenObservation = merge(tokenObservation, tokenData)
 	}
-	return tokenObservation, nil
+	return tokenDataObservations, nil
 }
 
-func merge(_ exectypes.TokenDataObservations, _ exectypes.TokenDataObservations) exectypes.TokenDataObservations {
-	return exectypes.TokenDataObservations{}
+func (t *CompositeTokenDataObserver) IsTokenSupported(
+	chainSelector cciptypes.ChainSelector,
+	token cciptypes.RampTokenAmount,
+) bool {
+	for _, ob := range t.observers {
+		if ob.IsTokenSupported(chainSelector, token) {
+			return true
+		}
+	}
+	return false
+}
+
+// initTokenDataObservations initializes the token data observations with empty token data, it asks the child observers
+// whether the token is supported or not. If token is supported and requires additional processing we set its state to
+// isReady=false. If token is noop (doesn't require offchain processing) we initialize it with empty
+// data and set IsReady=true.
+func (t *CompositeTokenDataObserver) initTokenDataObservations(
+	observations exectypes.MessageObservations,
+) exectypes.TokenDataObservations {
+	tokenObservation := make(exectypes.TokenDataObservations)
+	for chainSelector, observation := range observations {
+		tokenObservation[chainSelector] = make(map[cciptypes.SeqNum]exectypes.MessageTokenData)
+
+		for seq, message := range observation {
+			tokenData := make([]exectypes.TokenData, len(message.TokenAmounts))
+			for i, token := range message.TokenAmounts {
+				if !t.IsTokenSupported(chainSelector, token) {
+					// It means that none of the registers support that,
+					// we assume token doesn't require additional processing and skip it
+					tokenData[i] = exectypes.NewNoopTokenData()
```

Comment:
I thought we have to use `exectypes. NotSupportedTokenData` according to the doc comment above:

```golang
// * if token is not supported, TokenDataObserver must set the token data to exectypes.NotSupportedTokenData
```

Reply:
`compositeTokenDataObserver` works as an orchestrator here, it asks his children what tokens are supported and what not. Because we don't have an explicit list of tokens that don't require processing we implicitly get that by checking which tokens are supported by observers.

Therefore, we need to build some empty state and that's exactly what we do here by inserting `NewNoopTokenData()`. The comment I provided is true for any other `TokenDataObserver` that will be implemented in the future, but the composite has to cheat a bit here to initialize the empty state properly.